src/tvp/model.ts

```diff
@@ -0,0 +1,352 @@
+import { Emitter, Disposable } from 'vscode-jsonrpc'
+import { TextDocument, Position } from 'vscode-languageserver-types'
+
+import { TreeViewProvider } from './provider'
+import { TreeViewNode } from './domain'
+import { groupBy } from './utils'
+
+export interface NodeView {
+  underlying: TreeViewNode
+  level: number
+  expandable: boolean
+  expanded: boolean
+}
+
+export interface TreeModelUpdate {
+  root: NodeView
+  oldNodes: NodeView[]
+  newNodes: NodeView[]
+  focusEvent: boolean
+}
+
+export class Node {
+  private expanded: boolean
+  private children: Promise<Node[]> | undefined
+  private lastTouch = Date.now()
+  private defunct = false
+
+  constructor(
+    public viewNode: TreeViewNode,
+    public level: number,
+    private provider: TreeViewProvider,
+    private emitter: Emitter<TreeModelUpdate>
+  ) {
+    this.expanded = false
+  }
+
+  public expandable(): boolean {
+    return this.viewNode.collapseState !== undefined
+  }
+
+  public isExpanded(): boolean {
+    return this.expanded
+  }
+
+  public async expand(): Promise<boolean> {
+    if (this.defunct) return false
+
+    this.provider.sendTreeNodeVisibilityNotification(this.viewNode.nodeUri, false)
+    this.expanded = true
+    const curState = await this.touch()
+    await this.getChildren()
+    const newNodes = await this.collectVisibleNodes()
+    if (curState === this.lastTouch && !this.defunct) {
+      this.emitter.fire({root: this.makeView(), oldNodes: [this.makeView()], newNodes, focusEvent: false})
+      return true
+    } else {
+      return false
+    }
+  }
+
+  public async collapse(): Promise<boolean> {
+    if (this.defunct) return false
+
+    this.provider.sendTreeNodeVisibilityNotification(this.viewNode.nodeUri, true)
+    const curState = await this.touch()
+    const oldNodes = await this.collectVisibleNodes()
+    this.expanded = false
+    if (curState === this.lastTouch && !this.defunct) {
+      this.emitter.fire({root: this.makeView(), oldNodes, newNodes: [this.makeView()], focusEvent: false})
+      return true
+    } else {
+      return false
+    }
+  }
+
+  public async refreshSubtree(viewNode: TreeViewNode | undefined): Promise<void> {
+    async function findOpenNodes(acc: string[][], curPath: string[], node: Node): Promise<string[][]> {
+      if (node.isExpanded()) {
+        acc.push(curPath)
+        const children = await node.getChildren()
+        return children.reduce((zP, child) => {
```

Comment:
Can we use something more explicit that `zP`?

Reply:
I think that such names are quite natural for these variables in scala world.

`z` is standard name for accumulator of `foldLeft`, `foldRight` or `reduce`. Examples are [foldRight](https://github.com/scala/scala/blob/v2.13.1/src/library/scala/collection/immutable/List.scala#L320) or [foldLeft](https://github.com/scala/scala/blob/v2.12.10/src/library/scala/collection/LinearSeqOptimized.scala#L122).

Last capital letter is usually effect which is used somehow in current function or variables. This is used, for example, in [Monix](https://monix.io/api/3.0/monix/reactive/Observable.html) where a lot of methods have such naming: findL, firstL, completedL, comletedF, bracketF and so on. Cats-effect use [the same](https://typelevel.org/cats-effect/api/cats/effect/Concurrent.html).

So `zP` just means accumulator of function `reduce` wrapped inside `Promise`.

If you are disagree then I will rename.

Reply:
Understood. Thanks for the explanation. I'm fine with leaving them then.